aws/rust-runtime/aws-auth-providers/src/profile/repr.rs

```diff
@@ -0,0 +1,350 @@
+    // the profile must define either a `CredentialsSource` or a concrete set of access keys
+    let profile = match profile.get(role::CREDENTIAL_SOURCE) {
+        Some(source) => BaseProvider::NamedSource(source),
+        None => BaseProvider::AccessKey(static_creds_from_profile(profile)?),
+    };
+    Ok(profile)
+}
+
+enum NextProfile<'a> {
+    SelfReference,
+    Named(&'a str),
+}
+
+fn chain_provider(profile: &Profile) -> Option<Result<(RoleArn, NextProfile), ProfileFileError>> {
+    let role_provider = role_arn_from_profile(&profile)?;
+    let (source_profile, credential_source) = (
+        profile.get(role::SOURCE_PROFILE),
+        profile.get(role::CREDENTIAL_SOURCE),
+    );
+    let profile = match (source_profile, credential_source) {
+        (Some(_), Some(_)) => Err(ProfileFileError::InvalidCredentialSource {
+            profile: profile.name().to_string(),
+            message: "profile contained both source_profile and credential_source. \
+                Only one or the other can be defined"
+                .into(),
+        }),
+        (None, None) => Err(ProfileFileError::InvalidCredentialSource {
+            profile: profile.name().to_string(),
+            message:
+                "profile must contain source_profile or credentials_source but neither were defined"
+                    .into(),
+        }),
+        (Some(source_profile), None) if source_profile == profile.name() => {
+            Ok((role_provider, NextProfile::SelfReference))
+        }
+
+        (Some(source_profile), None) => Ok((role_provider, NextProfile::Named(source_profile))),
+        // we want to loop back into this profile and pick up the credential source
+        (None, Some(_credential_source)) => Ok((role_provider, NextProfile::SelfReference)),
+    };
+    Some(profile)
+}
+
+fn role_arn_from_profile(profile: &Profile) -> Option<RoleArn> {
+    let role_arn = profile.get(role::ROLE_ARN)?;
+    let session_name = profile.get(role::SESSION_NAME);
+    let external_id = profile.get(role::EXTERNAL_ID);
+    Some(RoleArn {
+        role_arn,
+        session_name,
+        external_id,
+    })
+}
+
+/// Load static credentials from a profile
+///
+/// Example:
+/// ```ini
+/// [profile B]
+/// aws_access_key_id = abc123
+/// aws_secret_access_key = def456
+/// ```
+fn static_creds_from_profile(profile: &Profile) -> Result<Credentials, ProfileFileError> {
+    use static_credentials::*;
+    let access_key = profile.get(AWS_ACCESS_KEY_ID);
+    let secret_key = profile.get(AWS_SECRET_ACCESS_KEY);
+    let session_token = profile.get(AWS_SESSION_TOKEN);
+    if let (None, None, None) = (access_key, secret_key, session_token) {
+        return Err(ProfileFileError::MissingCredentialSource {
+            profile: profile.name().to_string(),
+            message: "expected `aws_access_key_id` and `aws_secret_access_key` to be defined"
+                .into(),
+        });
+    }
```

Comment:
This error message will be confusing if it's the session token that's missing. Although, from reading the test cases, I'm not sure if that case is possible?

Reply:
the session token is optional so having that missing is never an error